lms/services/event.py

```diff
@@ -0,0 +1,33 @@
+from functools import lru_cache
+
+from sqlalchemy.orm import Session
+
+from lms.models import Event, EventData, EventType, EventUser
+
+
+class EventService:
+    def __init__(self, db: Session):
+        self._db = db
+
+    def insert_event(
+        self, type_: EventType, user: None, lti_roles=None, data=None, **kwargs
+    ):
+
+        event = Event(type_id=self._get_type_pk(type_), **kwargs)
+        self._db.add(event)
+
+        if user:
+            for lti_role in lti_roles or [None]:
+                self._db.add(EventUser(event=event, user=user, lti_role=lti_role))
+
+        if data:
+            print(data)
+            self._db.add(EventData(event=event, data=data))
+
+    @lru_cache
+    def _get_type_pk(self, type_: EventType.Type) -> int:
```

Comment:
Need to return the ID, otherwise the models from other sessions would get cached and break.

Reply:
I don't really understand this. Can't you put in the string and let SQLAlchemy work it out?

Reply:
The table main table is using `event_id` and then we have a table with a PK and the type name. 

The idea is to save a bit space there in the long run storing only the FK on the main table.

To do this without this trick we'll either make a query manually to get the PK of the type every time of use some join on the insert. I reckon this is small enough to be worth it.

Reply:
Right gotcha. 

I think you might want to put a fixed number on the `lru_cache` here, as otherwise it can end up causing memory leaks when used in a method with `self`. I'm not going to pretend to understand why, but lint told me so.